### DISCUSSION TITLE
export html with widgets

### DISCUSSION THREAD

**Discussion by edouardbutaye** [pyvista/pyvista#5315] [Q&A]
Hi,
I looked for similar discussions but couldn't find any answers. 
I would like to know if there is a command to export an interactive pyvista plot to html format (or another one).
For example (slight modification of the example https://docs.pyvista.org/version/stable/api/plotting/_autosummary/pyvista.Plotter.add_slider_widget.html#pyvista.Plotter.add_slider_widget) :
```

import pyvista as pv
pl = pv.Plotter()

def create_mesh(value):
      res = int(value)
      sphere = pv.Sphere(
      phi_resolution=res, theta_resolution=res)
      pl.add_mesh(sphere, name="sphere", show_edges=True)

slider = pl.add_slider_widget(
  create_mesh,
  [5, 100],
  title="Resolution",
  title_opacity=0.5,
  title_color="red",
  fmt="%0.9f",
  title_height=0.08,
)

pl.export_html()
pl.show()
```

Sincerely,
Edouard Butaye

**Comment by olmasset** [comment]
I was wondering the same thing for a moving colormap. For example, simply adding 

`pltr.export_html('movie.html') `

at the end of the example "https://docs.pyvista.org/version/stable/examples/02-plot/moving_cmap.html#create-a-gif-movie-of-a-static-object-with-a-moving-colormap" will not work. It will create an html of the static object.


```
import numpy as np

import pyvista as pv


# A spherical curve
def scurve(t):
    alpha = np.pi / 2 - (np.pi / 2 - 0.44) * np.cos(3 * t)
    beta = t + 0.44 * np.sin(6 * t)
    return np.array(
        [np.sin(alpha) * np.cos(beta), np.sin(alpha) * np.sin(beta), np.cos(alpha)]
    )


# Hopf fiber
def hopf_fiber(p, phi):
    return np.array(
        [
            (1 + p[2]) * np.cos(phi),
            p[0] * np.sin(phi) - p[1] * np.cos(phi),
            p[0] * np.cos(phi) + p[1] * np.sin(phi),
            (1 + p[2]) * np.sin(phi),
        ]
    ) / np.sqrt(2 * (1 + p[2]))


# Stereographic projection
def stereo_proj(q):
    return q[0:3] / (1 - q[3])


# Parameterization of the Hopf torus
def hopf_torus(t, phi):
    return stereo_proj(hopf_fiber(scurve(t), phi))


# Create the mesh
angle_u = np.linspace(-np.pi, np.pi, 400)
angle_v = np.linspace(0, np.pi, 200)
u, v = np.meshgrid(angle_u, angle_v)
x, y, z = hopf_torus(u, v)
grid = pv.StructuredGrid(x, y, z)
mesh = grid.extract_geometry().clean(tolerance=1e-6)

# Distances normalized to [0, 2*pi]
dists = np.linalg.norm(mesh.points, axis=1)
dists = 2 * np.pi * (dists - dists.min()) / (dists.max() - dists.min())

mesh["distances"] = np.sin(dists)

# Make the movie
pltr = pv.Plotter(window_size=[512, 512])
pltr.set_focus([0, 0, 0])
pltr.set_position([40, 0, 0])
pltr.add_mesh(
    mesh,
    scalars="distances",
    smooth_shading=True,
    specular=1,
    cmap="nipy_spectral",
    show_scalar_bar=False,
)
pltr.open_gif("Hopf_torus.gif")

for t in np.linspace(0, 2 * np.pi, 60, endpoint=False):
    mesh["distances"] = np.sin(dists - t)
    pltr.write_frame()
pltr.export_html("movie.html")
pltr.show()
```

**Comment by banesullivan** [comment]
We cannot export the embedded widgets unfortunately -- these require a live Python kernel and will not serialize to VTK.js

**Reply by edouardbutaye** [reply]
Oh okay, thank you for your answer!

**Reply by olmasset** [reply]
Thank you for the reply. I am trying to do it with dash-vtk. Hoping to find something similar to the pyvista.write_frame() formalism.

**Reply by banesullivan** [reply]
@jourdain may have tips or updates on the latest capabilities of VTK.js and trame-vtk (and he's the author of dash-vtk 😉)

**Reply by jourdain** [reply]
dash-vtk does not have 3d widgets and also requires some server side. Can you explain what you are trying to have as the client side only?

**Reply by olmasset** [reply]
Of course! I am trying to get a moving colormap, juste like the one displayed in the example (https://docs.pyvista.org/version/stable/examples/02-plot/moving_cmap#sphx-glr-examples-02-plot-moving-cmap-py), and a slider that will change some parameters. 

I did a dash app in which I display a VTK rendering using `dash_vtk.View ` and I have some dash core component slider  `dcc.Slider`. 
So the part I am missing is the "moving" part. What in the pyvista example is contained in the loop :

```
# Make the movie
pltr = pv.Plotter(window_size=[512, 512])
pltr.set_focus([0, 0, 0])
pltr.set_position([40, 0, 0])
pltr.add_mesh(
    mesh,
    scalars="distances",
    smooth_shading=True,
    specular=1,
    cmap="nipy_spectral",
    show_scalar_bar=False,
)
pltr.open_gif("Hopf_torus.gif")

for t in np.linspace(0, 2 * np.pi, 60, endpoint=False):
    mesh["distances"] = np.sin(dists - t)
    pltr.write_frame()

pltr.show()
```

**Reply by jourdain** [reply]
In the current state, you can not do that. I mean you can do it with a server by changing the data array like you are doing but not in a plain offline HTML file. Technically we could create a trame app that does that on the client side only but additional work will be require to allow to export a trame app as a static web bundle.

**Reply by olmasset** [reply]
Thank you for the reply.

**Reply by tgratek** [reply]
Hi, I am using PyVista for a project, and am curious whether the export to html problem here also applies on PyVista + Trame applications that uses Trame's interactive menu options.

For example: https://tutorial.pyvista.org/tutorial/09_trame/b_trame_actor_color.html has simply a dropdown option to select a color in the page layout. When exported to html, we can interact with the object, but the option to change its color disappears. 

Are there workarounds for this?

**Reply by jourdain** [reply]
> Technically we could create a trame app that does that on the client side only but additional work will be require to allow to export a trame app as a static web bundle.

No workaround to date if you involve trame. It does require a server by default.
